examples/beacon-api-sidecar-fetcher/src/main.rs

```diff
@@ -0,0 +1,297 @@
+
+    #[error("There was an error grabbing the blob from the tx pool: {0}")]
+    TransactionPoolError(BlobStoreError),
+
+    #[error("400: {0}")]
+    InvalidBlockID(String),
+
+    #[error("404: {0}")]
+    BlockNotFound(String),
+
+    #[error("500: {0}")]
+    InternalError(String),
+
+    #[error("Network error: {0}")]
+    NetworkError(String),
+
+    #[error("Data parsing error: {0}")]
+    DeserializationError(String),
+
+    #[error("{0} Error: {1}")]
+    UnknownError(u16, String),
+}
+
+/// tracks the futures associated with retrieving blob data from the beacon client
+type BeaconFutures =
+    FuturesUnordered<Pin<Box<dyn Future<Output = Result<Vec<BlobTransaction>, SideCarError>>>>>;
+
+/// Wrapper struct for CanonStateNotifications
+pub struct MinedSidecarStream<St, P>
+where
+    St: Stream<Item = CanonStateNotification> + Send + Unpin + 'static,
+{
+    events: St,
+    pool: P,
+    beacon_config: BeaconSidecarConfig,
+    client: reqwest::Client,
+    pending_requests: BeaconFutures,
+    queued_actions: VecDeque<BlobTransaction>,
+}
+
+/// First checks if the Blobtranscation for a given EIP4844 is stored locally, if not attempts to
+/// retrieve it from the CL Layer
+impl<St, P> Stream for MinedSidecarStream<St, P>
+where
+    St: Stream<Item = CanonStateNotification> + Send + Unpin + 'static,
+    P: TransactionPoolExt + Unpin + 'static,
+{
+    type Item = Result<BlobTransaction, SideCarError>;
+
+    /// Attempt to pull the next BlobTransaction from the stream.
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this: &mut MinedSidecarStream<St, P> = self.get_mut();
+
+        if let Some(mined_sidecar) = this.queued_actions.pop_front() {
+            return Poll::Ready(Some(Ok(mined_sidecar)));
+        }
+
+        // Request locally first, otherwise request from CL
+        loop {
+            // Check if any pending reqwests are ready and append to buffer
+            while let Poll::Ready(Some(pending_result)) = this.pending_requests.poll_next_unpin(cx)
+            {
+                match pending_result {
+                    Ok(mined_sidecars) => {
+                        for sidecar in mined_sidecars {
+                            this.queued_actions.push_back(sidecar);
+                        }
+                    }
+                    Err(err) => return Poll::Ready(Some(Err(err))),
+                }
+            }
+
+            while let Poll::Ready(Some(notification)) = this.events.poll_next_unpin(cx) {
+                {
+                    let mut all_blobs_available = true;
+                    let mut actions_to_queue: Vec<BlobTransaction> = Vec::new();
+
+                    // Collect only the EIP4844 transcations.
+                    let txs: Vec<_> = notification
+                        .tip()
+                        .transactions()
+                        .filter(|tx: &&reth::primitives::TransactionSigned| tx.is_eip4844())
```

Comment:
this needs to be a bit more granular, because this can
a) include multiple blocks
b) can be a reorged block(s)

Reply:
I added granularity matching on the notifications. However, I have a couple of questions:

1. Should we handle commits differently during a reorg? Specifically, should we send the hashes for the old block and then execute the existing logic for new blocks? If so, I will extract this into a new function.

2. How do we determine the level of granularity here? Do we use new.blocks()?

Reply:
> Specifically, should we send the hashes for the old block and then execute the existing logic for new blocks?

i think so, yes

> How do we determine the level of granularity here? Do we use new.blocks()?

unclear wym by granularity, but every committed block needs to be processed